src/server/game/AI/CreatureAI.cpp

```diff
@@ -271,6 +271,112 @@ bool CreatureAI::_EnterEvadeMode()
+    }
+    bool hadBoundsWarning = false;
+    Q.push({ 0,0 });
+    while (!Q.empty())
+    {
+        coordinate front = Q.front();
+        bool hasOutOfBoundsNeighbor = false;
+        for (coordinate off : std::list<coordinate>({ {1, 0}, { 0,1 }, { -1,0 }, { 0,-1 } }))
+        {
+            coordinate next(front.first + off.first, front.second + off.second);
+            if (next.first > BOUNDARY_VISUALIZE_FAILSAFE_LIMIT || next.first < -BOUNDARY_VISUALIZE_FAILSAFE_LIMIT || next.second > BOUNDARY_VISUALIZE_FAILSAFE_LIMIT || next.second < -BOUNDARY_VISUALIZE_FAILSAFE_LIMIT)
+            {
+                if (!hadBoundsWarning)
+                {
+                    hadBoundsWarning = true;
+                    me->Talk("VisualizeBoundary warning: Reached fail-safe flood boundary - check is your boundary is unbounded!", CHAT_MSG_MONSTER_YELL, LANG_UNIVERSAL, 50000.0f, nullptr);
+                }
+                continue;
+            }
+            if (alreadyChecked.find(next) == alreadyChecked.end()) // never check a coordinate twice
+            {
+                Position nextPos(meX + next.first*BOUNDARY_VISUALIZE_STEP_SIZE, meY + next.second*BOUNDARY_VISUALIZE_STEP_SIZE, meZ);
+                if (CheckBoundary(&nextPos))
+                    Q.push(next);
+                else
+                {
+                    outOfBounds.insert(next);
+                    hasOutOfBoundsNeighbor = true;
+                }
+                alreadyChecked.insert(next);
+            }
+            else
+                if (outOfBounds.find(next) != outOfBounds.end())
+                    hasOutOfBoundsNeighbor = true;
+        }
+        if(fill || hasOutOfBoundsNeighbor)
+            if (TempSummon* point = owner->SummonCreature(BOUNDARY_VISUALIZE_CREATURE, Position(meX + front.first*BOUNDARY_VISUALIZE_STEP_SIZE, meY + front.second*BOUNDARY_VISUALIZE_STEP_SIZE, spawnZ), TEMPSUMMON_TIMED_DESPAWN, duration * IN_MILLISECONDS))
+            {
+                point->SetObjectScale(BOUNDARY_VISUALIZE_CREATURE_SCALE);
+                point->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_IMMUNE_TO_PC | UNIT_FLAG_STUNNED | UNIT_FLAG_IMMUNE_TO_NPC);
+                if (!hasOutOfBoundsNeighbor)
+                    point->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NOT_SELECTABLE);
+            }
+        Q.pop();
+    }
+}
+
+bool CreatureAI::CheckBoundary(Position* who) const
+{
+    if (!who)
+        who = me;
+
+    if (_boundary)
+        for (CreatureBoundary::const_iterator it = _boundary->begin(); it != _boundary->end(); ++it)
+            if ((*it) && !(*it)->IsWithinBoundary(who))
+                return false;
+
+    return true;
+}
+
+bool CreatureAI::CheckInRoom()
+{
+    if (CheckBoundary())
+        return true;
+
+    switch (GetBoundaryEvadeType())
+    {
+        case EVADE_TYPE_RETURN:
+            EnterEvadeMode();
+            break;
+        case EVADE_TYPE_RESPAWN:
+            me->DespawnOrUnsummon();
+            me->SetRespawnTime(15);
```

Comment:
On SOO (Siege of Orgrimmar), all boss give "despawn" regardless of which way the evade occurred. On Retail (WoD), all boss on Icecrown Citadel give despawn for all evade conditions too.